### PR TITLE
VEP 115: Add PCIe NUMA topology awareness VEP

### PR DESCRIPTION
**Pull request by mresvanis**
### VEP Metadata

**Tracking issue**:  https://github.com/kubevirt/enhancements/issues/115
**SIG label**:  /sig compute

### What this PR does
When `guestMappingPassthrough` is set any PCIe GPU and host devices with NUMA affinity information will be placed to a PCI controller hierarchy matching their NUMA node affinity (i.e. not directly on the default PCI root). The devices that don't have any NUMA alignment will be assigned to the default PCI root. Any error during this process will result to all devices falling back to their assignment under the default PCI root.

### CODE REVIEW

File: veps/sig-compute/pcie-numa-topology-awareness.md [hunk 1]
```diff
@@ -0,0 +1,544 @@
...
+
+Current limitations:
+- Guest VMs lack visibility into GPU and NIC NUMA placement and device locality.
+- AI frameworks cannot optimize communication topology for multi-GPU distributed model training.
+- Cross-NUMA memory traffic degrades performance despite correct NUMA-aware device allocation.
+
+## Goals
+
+- Add a new feature gate `PCINUMAAwareTopology`.
+- When the afore-mentioned feature gate and `guestMappingPassthrough` are enabled, mirror PCIe
```
[the remark this answers is not in the document]
Comment (fanzhangio) [reply veps/sig-compute/pcie-numa-topology-awareness.md]: Thanks @lyarwood for your feedback!
Our current design is based on the understanding that PCIe NUMA passthrough is inherently coupled with CPU NUMA passthrough. In practice, enabling PCIe NUMA passthrough without aligning the guest’s CPU topology is not feasible. Therefore, the workflow we designed assumes that:

- If users want to enable PCIe NUMA awareness, they must explicitly enable the feature gate and set `guestMappingPassthrough: true`
- If users only want CPU NUMA passthrough, the current behavior continues to work as expected, as long as the feature gate remains disabled.
 
That said, you bring up an excellent point: when the feature gate graduates, enabling it could inadvertently alter behavior for users already relying on `guestMappingPassthrough`. Introducing an explicit API field early on is a more robust and future-proof solution, as it avoids implicit behavior shifts.

I agree with your suggestion. We can introduce a dedicated policy field to clearly express PCIe NUMA passthrough intent. One possible design might look like this:
```yaml
spec:
  domain:
    devices:
      pciNumaAwareTopology: true
```

[the remark this answers is not in the document]
Comment (vladikr) [reply veps/sig-compute/pcie-numa-topology-awareness.md]: @lyarwood, I'm wondering if you could explain why we would want another tunable?
Are there users who don't want to expose the NUMA affinity of the passed-through devices?
So far, we have been avoiding constructing PCI controllers, but with this VEP, we have a path forward. 
I think that constructing these controller hierarchies for NUMA passthrough is a first step, but later we will need to find a way to advertise the passed devices' NUMA affinity with dedicated CPUs as well  - hopefully with DRA integration, this will be easier

if we must have a new api field, then I would vote for
 ```
spec:
  domain:
    cpu:
      numa:
        guestMappingPassthrough: {}
        pciNumaAwareTopology: {}
```

However, I don't know if it's needed.

[the remark this answers is not in the document]
Comment (lyarwood) [reply veps/sig-compute/pcie-numa-topology-awareness.md]: @vladikr guest ABI, ultimately to implement this we are going to rewire PCI device topology right? So without an additional configurable all users of `guestMappingPassthrough` will see this enabled in their guests once the FG graduates or is enabled. That's why I'm suggesting another configurable.

[the remark this answers is not in the document]
Comment (vladikr) [reply veps/sig-compute/pcie-numa-topology-awareness.md]: @lyarwood I still fail to see a reason for a VMI level api to enable to disable this. For me, this will be just another tunable that the user will forget to set.
From a live migration point of view, these additional controllers will come into play only if we have a passthrough device - for most, we don't allow live migration. So there is no upgrade path for these VMs. The only exception is network SRIOV, which in some cases we unplug and re-plug the device on the target - I'm not sure here @EdDev @orelmisan fyi
I think there shouldn't be any effect since we are reusing the same domain xml from the running domain and not creating a new one.

In the static case, when a stopped VM gets new controllers, we will need to evaluate this. I don't believe there is an issue with Linux, but I don't know about Windows guests. 
However, I don't believe we need a VM-level tunable; instead, we could have a cluster-wide setting.

File: veps/sig-compute/pcie-numa-topology-awareness.md [hunk 2]
```diff
@@ -0,0 +1,561 @@
...
+  host device NUMA topology in the guest.
+
+## Non Goals
+
+- This VEP focuses solely on ensuring optimal guest PCIe topology generation
+  when devices are locally available. It does not address changes to Kubernetes
+  device/resource allocation for guaranteeing NUMA-aligned device allocation.
+  Dynamic Resource Allocation (DRA) will help address that in the near future.
+- Hot-plug support, as `pcie-expander-bus` controllers are [not hot-pluggable](https://github.com/qemu/qemu/blob/711a1ddf899bef577907a10db77475c8834da52f/docs/pcie.txt#L222-L228).
```
Comment (Aseeef) [inline veps/sig-compute/pcie-numa-topology-awareness.md]: pcie-expander-buses are not hot pluggable as stated. But what about devices connected through the pcie-root-port (which if I understand do implement hot plugging support)?

Basically when we say "no hotplug support" is a non-goal does that include the downstream devices that would be connecting into the pcie-root-port?

File: veps/sig-compute/pcie-numa-topology-awareness.md [hunk 3]
```diff
@@ -0,0 +1,561 @@
...
+          - deviceName: rdma/ib_device
+            name: ib1
+        resources:
+          requests:
+            memory: 8Gi
+            nvidia.com/gpu: 4
+            rdma/ib_device: 2
+```
+
+## Implementation Roadmap
```
Comment (Aseeef) [inline veps/sig-compute/pcie-numa-topology-awareness.md]: I noticed graduation requirements put mediated devices in a separate stage and looking at that I think we should explicitly mention at what stage here we intend to do the implementation for VFs and mdevs.

File: veps/sig-compute/pcie-numa-topology-awareness.md [hunk 4]
```diff
@@ -0,0 +1,561 @@
...
+</devices>
+```
+
+## Scalability
+
+The solution should scale to:
+
+- Large clusters with hundreds of NUMA-enabled nodes.
+- Multiple PCIe devices per NUMA node (up to 14+ devices as shown in examples).
+- Complex multi-socket systems with 4+ NUMA nodes.
```
Comment (Aseeef) [inline veps/sig-compute/pcie-numa-topology-awareness.md]: About the "14+ devices" and "4+ NUMA nodes" as mentioned: is there a fundamental limitation for why we can't or shouldn't support systems with more devices per NUMA node or systems with more NUMA nodes besides the 255 devices hard-limit on the root-complex (pcie.0)?

Meaning for example, would a system with 8 NUMA nodes with 2 devices each (16 total) still not be valid?

Comment (mresvanis) [reply veps/sig-compute/pcie-numa-topology-awareness.md]: I apologize, I will rephrase these 2 sentences to better reflect their intention.  The `14+ devices` and `4+ NUMA nodes`  are there to show that this VEP's proposed solution should cover systems with more that 4 NUMA nodes and more than 14 devices per NUMA node. 

> Meaning for example, would a system with 8 NUMA nodes with 2 devices each (16 total) still not be valid?

That system should be properly supported by this VEP's proposed solution.

File: veps/sig-compute/pcie-numa-topology-awareness.md (hunk 3 above)
Comment (Aseeef) [reply veps/sig-compute/pcie-numa-topology-awareness.md]: OK, I see. So this implementation road map is specifically referring to the "Alpha" part of this later section, right?
```
## Graduation Requirements

### Alpha
- Basic NUMA topology awareness implemented
- Unit tests and basic integration tests
- Documentation for feature usage

### Beta
- Support for mediated devices and VFs
- Performance validation with real AI workloads
- Comprehensive testing coverage including edge cases
- Integration with DRA (Dynamic Resource Allocation)

### GA
- Production-ready performance and reliability
- Full documentation and operational guides
```

I was getting confused how this section corresponded to this later section.

File: veps/sig-compute/pcie-numa-topology-awareness.md [hunk 5]
```diff
@@ -0,0 +1,560 @@
...
+
+- Add a new feature gate `PCINUMAAwareTopology`.
+- When the afore-mentioned feature gate and `guestMappingPassthrough` are enabled, mirror PCIe
+  host device NUMA topology in the guest.
+
+## Non Goals
+
+- Changes in Kubernetes device/resource allocation to guarantee NUMA-aligned device assignment.
+  This VEP focuses solely on guest PCIe topology generation.
+- Hot-plug support, as `pcie-expander-bus` controllers are [not hot-pluggable](https://github.com/qemu/qemu/blob/711a1ddf899bef577907a10db77475c8834da52f/docs/pcie.txt#L222-L228).
```
[the remark this answers is not in the document]
Comment (xpivarc) [reply veps/sig-compute/pcie-numa-topology-awareness.md]: @orelmisan @EdDev cc

File: veps/sig-compute/pcie-numa-topology-awareness.md (hunk 3 above)
Comment (mresvanis) [reply veps/sig-compute/pcie-numa-topology-awareness.md]: That's a very good point, I now realized that we can improve this and be clear about the milestones for each version/implementation. I'll update the doc and ping you back.

File: veps/sig-compute/pcie-numa-topology-awareness.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (EdDev) [reply veps/sig-compute/pcie-numa-topology-awareness.md]: I would prefer the non-goals to be generic as well, not specific and without too much details.
At this level, we are interested to know if hotplug of hostdevices (or devices in general?) is supported or not.
The implications regarding SR-IOV is dependent on the ability to migrate + hotplug. If we cannot migrate and/or hotplug the device, SR-IOV is not hotpluggable, like many other hotplug functionalities.
Therefore, I would mention that any dependent migration functionality, like hotplug, is not supported.

> **SR-IOV Virtual Functions (VFs)**: VFs inherit NUMA affinity from their parent Physical Function (PF). KubeVirt resolves the parent PF by reading the `physfn` symlink at `/sys/bus/pci/devices/<VF_BDF>/physfn`, which points to the parent PF's sysfs directory. The VF is then assigned to the same NUMA node and `pcie-expander-bus` controller as its parent PF. This ensures that all VFs from the same PF maintain locality with their shared hardware resources, enabling optimal performance for SR-IOV workloads.

I do not think the virt-launcher has access to the PF. Only the VF is passed to the pod and the PF is touched by the CNI which has access to the node namespace.

File: veps/sig-compute/pcie-numa-topology-awareness.md [hunk 6]
```diff
@@ -0,0 +1,561 @@
...
+- Add a new feature gate `PCINUMAAwareTopology`.
+- When the afore-mentioned feature gate and `guestMappingPassthrough` are enabled, mirror PCIe
+  host device NUMA topology in the guest.
+
+## Non Goals
+
+- This VEP focuses solely on ensuring optimal guest PCIe topology generation
+  when devices are locally available. It does not address changes to Kubernetes
+  device/resource allocation for guaranteeing NUMA-aligned device allocation.
+  Dynamic Resource Allocation (DRA) will help address that in the near future.
```
Comment (EdDev) [inline veps/sig-compute/pcie-numa-topology-awareness.md]: I do not understand this part. Without it, how can we provide a solution that works?
AFAIU the NUMA topology awareness is dependent on the availability of other resource managers to assign correct devices into the pod, such that Kubevirt will be able to continue and assign those to the domain.

If that initial step is not something we can assure, the feature is not complete.
I would expect that Kubevirt is able to request this in a way from the resource manager and be assured it is aligned.

To elaborate a bit more, if this is dependent on an external solution that does this correct assignment, I would like to understand how it interacts with Kubevirt. As a community project, we should at the minimum give some guidelines and possibly references to existing solutions which can make this happen.

We'll also need to test this, therefore it is also a practical scenario which needs attention.

File: veps/sig-compute/pcie-numa-topology-awareness.md [hunk 7]
```diff
@@ -0,0 +1,561 @@
...
+                                            ---------------           ---------------
+                                            | PCIe Device |           | PCIe Device |
+                                            |  (GPU/NIC)  |           |  (GPU/NIC)  |
+                                            ---------------           ---------------
+```
+
+### Scope
+
+- This feature is enabled only when:
+  - the feature gate `PCINUMAAwareTopology` is enabled and
```
Comment (EdDev) [inline veps/sig-compute/pcie-numa-topology-awareness.md]: This is irrelevant for the feature itself. The FG protects the feature lifecycle and is not used to control it as a operational knob. The design should focus on graduation and keep the lifecycle in the relevant section.
If you need to control this feature operation, a formal API is needed.

File: veps/sig-compute/pcie-numa-topology-awareness.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (orelmisan) [reply veps/sig-compute/pcie-numa-topology-awareness.md]: The current suggestion is written in the present tense, but I haven't found any NUMA-aware logic for defining SR-IOV VFs in our current implementation.
I assume adding this logic is one of the intended goals of the PR?

Does the virt-launcher pod have enough privileges to read the relevant information from the operating system?

File: veps/sig-compute/pcie-numa-topology-awareness.md [hunk 8]
```diff
@@ -0,0 +1,561 @@
...
+
+1. Discovers the NUMA node for each passthrough PCIe device and if it matches any of the guest's NUMA nodes.
+2. Filters out devices without NUMA alignment to be placed on the default `pci.0` bus using the existing root slot assigner.
+3. Groups devices by NUMA node: `numaNode -> []api.HostDevice`.
+4. Creates `pcie-expander-bus` controllers with `<target busNr="X"><node>Y</node></target>` elements for each NUMA node.
+5. Creates one `pcie-root-port` controller per device under the respective `pcie-expander-bus` controller.
+6. Assigns device addresses to the `pcie-root-port` aligned with its NUMA node.
+7. Inserts the controllers created into the domain before adding the host devices.
+
+### Schema Changes
```
Comment (EdDev) [inline veps/sig-compute/pcie-numa-topology-awareness.md]: Is this in sync with https://gitlab.com/libvirt/libvirt-go-xml-module ?

File: veps/sig-compute/pcie-numa-topology-awareness.md [hunk 9]
```diff
@@ -0,0 +1,561 @@
...
+  expander bus, and bus numbers less than the specified value will be available to the next lower
+  expander-bus (or the root-bus if there are no lower expander buses). If you do not specify a busNumber,
+  libvirt will find the lowest existing busNumber in all other expander buses (or use 256 if there are
+  no others) and auto-assign the busNr of that found bus - 2, which provides one bus number for the
+  pci-expander-bus and one for the pci-bridge that is automatically attached to it (if you plan on adding
+  more pci-bridges to the hierarchy of the bus, you should manually set busNr to a lower value).
+
+### Hardware Utils Extension
+
+We will extend the `hardware` package with a function to retrieve a PCIe device's NUMA node
```
Comment (EdDev) [inline veps/sig-compute/pcie-numa-topology-awareness.md]: What package is this referring to? Can you provide a ref and/or overview of what is extended?

File: veps/sig-compute/pcie-numa-topology-awareness.md [hunk 10]
```diff
@@ -0,0 +1,561 @@
...
+The solution should scale to:
+
+- Large clusters with hundreds of NUMA-enabled nodes.
+- Multiple PCIe devices per NUMA node (up to 14+ devices as shown in examples).
+- Complex multi-socket systems with 4+ NUMA nodes.
+- Efficient controller allocation avoiding resource conflicts.
+
+## Update/Rollback Compatibility
+
+- Feature is opt-in via existing `spec.domain.cpu.numa.guestMappingPassthrough` flag.
```
Comment (EdDev) [inline veps/sig-compute/pcie-numa-topology-awareness.md]: What about compatibility with this opt-in setting to this new addition? It was discussed before that it will change the topology. Can you add that information here?

File: veps/sig-compute/pcie-numa-topology-awareness.md [hunk 11]
```diff
@@ -0,0 +1,561 @@
...
+- Existing VMs without NUMA specifications continue to work unchanged.
+- Devices reporting NUMA node `-1` fallback to the default `pci.0` layout.
+- Rollback gracefully disables NUMA-aware features without affecting running workloads.
```
Comment (EdDev) [inline veps/sig-compute/pcie-numa-topology-awareness.md]: What does "rollback" means in this context? Kubevirt is not supporting downgrading, therefore this may effect a single VM to disable the mapping. Was that the case referred here?

How will a guest OS behave if you change the scheme?

File: veps/sig-compute/pcie-numa-topology-awareness.md [hunk 12]
```diff
@@ -0,0 +1,561 @@
...
+
+## Functional Testing Approach
+
+- Unit tests: add cases under `virt-launcher` `virtwrapper` codebase verifying NUMA grouping, controller
+  generation, and fallback when NUMA is `-1`.
+- Integration tests: craft NUMA VMI functional tests (similar to the ones found in `tests/numa/numa.go`)
+  launching with multiple passthrough GPUs on different NUMA nodes; assert domain XML includes `pxb-pcie` `node='0/1'`
+  and devices placed under the correct root ports.
+- Provide a golden XML fixture akin to `numa-topo-vm-domain.xml` for regression checks.
```
Comment (EdDev) [inline veps/sig-compute/pcie-numa-topology-awareness.md]: Is this referring to unit tests?

**Review by EdDev — CHANGES_REQUESTED** [review]
Thank you for this interesting proposal.

The overall feature proposal is clear.
I am missing details on how this is going to play in the different kubevirt components and the API seems not to reach agreement yet.

I may not have all the context the other reviewers of this VEP had, therefore I apologize in advance.

File: veps/sig-compute/pcie-numa-topology-awareness.md (hunk 4 above)
Comment (mresvanis) [reply veps/sig-compute/pcie-numa-topology-awareness.md]: I updated the text to make this point clear. PTAL

File: veps/sig-compute/pcie-numa-topology-awareness.md (hunk 7 above)
Comment (mresvanis) [reply veps/sig-compute/pcie-numa-topology-awareness.md]: Great point, I removed the feature gate from the scope. PTAL

File: veps/sig-compute/pcie-numa-topology-awareness.md [hunk 13]
```diff
@@ -0,0 +1,561 @@
...
+- Feature is opt-in via existing `spec.domain.cpu.numa.guestMappingPassthrough` flag.
+- Existing VMs without NUMA specifications continue to work unchanged.
+- Devices reporting NUMA node `-1` fallback to the default `pci.0` layout.
+- Rollback gracefully disables NUMA-aware features without affecting running workloads.
+
+## Functional Testing Approach
+
+- Unit tests: add cases under `virt-launcher` `virtwrapper` codebase verifying NUMA grouping, controller
+  generation, and fallback when NUMA is `-1`.
+- Integration tests: craft NUMA VMI functional tests (similar to the ones found in `tests/numa/numa.go`)
```
[the remark this answers is not in the document]
Comment (mresvanis) [reply veps/sig-compute/pcie-numa-topology-awareness.md]: Good point, I have updated all text referring to "integration tests" to use "end-to-end test" instead. PTAL

File: veps/sig-compute/pcie-numa-topology-awareness.md [hunk 14]
```diff
@@ -0,0 +1,561 @@
...
+  launching with multiple passthrough GPUs on different NUMA nodes; assert domain XML includes `pxb-pcie` `node='0/1'`
+  and devices placed under the correct root ports.
+- Provide a golden XML fixture akin to `numa-topo-vm-domain.xml` for regression checks.
+- Performance benchmarks comparing NUMA-aware vs NUMA-unaware configurations.
```
[the remark this answers is not in the document]
Comment (mresvanis) [reply veps/sig-compute/pcie-numa-topology-awareness.md]: Since we are not ready to answer these questions I opted to remove this point from that section. I am also not confident that these benchmarks would be useful. Does this make sense?

File: veps/sig-compute/115-pcie-numa-topology-awareness/pcie-numa-topology-awareness.md [hunk 15]
```diff
@@ -0,0 +1,561 @@
...
+  expander bus, and bus numbers less than the specified value will be available to the next lower
+  expander-bus (or the root-bus if there are no lower expander buses). If you do not specify a busNumber,
+  libvirt will find the lowest existing busNumber in all other expander buses (or use 256 if there are
+  no others) and auto-assign the busNr of that found bus - 2, which provides one bus number for the
+  pci-expander-bus and one for the pci-bridge that is automatically attached to it (if you plan on adding
+  more pci-bridges to the hierarchy of the bus, you should manually set busNr to a lower value).
+
+### Hardware Utils Extension
+
+We will extend the `hardware` package with a function to retrieve a PCIe device's NUMA node
```
Comment (mresvanis) [reply veps/sig-compute/115-pcie-numa-topology-awareness/pcie-numa-topology-awareness.md]: Great catch, I added a link to the [respective](https://github.com/kubevirt/kubevirt/tree/7d7c72f1a0ec134175a750d1d89d261fd6138823/pkg/util/hardware) package and a sample implementation. PTAL

File: veps/sig-compute/115-pcie-numa-topology-awareness/pcie-numa-topology-awareness.md [hunk 16]
```diff
@@ -0,0 +1,561 @@
...
+  - the feature gate `PCINUMAAwareTopology` is enabled and
+  - the virtual machine's `spec.domain.cpu.numa.guestMappingPassthrough` is set.
+- It applies only to passthrough PCIe host devices (GPUs, IB/RoCE NICs, SR-IOV PF/VFs, DRA-backed host devices).
+- Mediated devices (mdevs) follow the parent physical PCI device's NUMA node when available. Mediated devices
+  should be backed by Virtual Functions (VFs) using VFIO passthrough.
+- Devices without NUMA affinity are placed on the default `pci.0` bus and report NUMA node `-1` (absent).
+
+### Device NUMA Affinity Discovery
+
+KubeVirt discovers device NUMA affinity by reading `/sys/bus/pci/devices/<BDF>/numa_node`
```
[the remark this answers is not in the document]
Comment (mresvanis) [reply veps/sig-compute/115-pcie-numa-topology-awareness/pcie-numa-topology-awareness.md]: > Which component is expected to read it and at what stage?
 
Very good question, I completely missed adding the relevant details. What we have already tested (PoC-ed) is gathering the NUMA node of the host devices via the `virt-launcher`'s VMI Spec to libvirt domain XML [conversion](https://github.com/kubevirt/kubevirt/blob/0358eea1e1b9faf2717700b7b0599268e8488706/pkg/virt-launcher/virtwrap/converter/converter.go#L1159). In the same flow we can also setup the respective PCIe controllers to reflect the devices' NUMA affinity when that is available. I added these details to the VEP. PTAL

> In addition, resolving parent devices and attempting to query them may also hit the permission and visibility challenge. E.g. with SR-IOV, while VFs are assigned to the pod, their parent device (PF) is not.

> There may be a need to run a PoC to check all these points.
> It will be valuable for this information to be explicitly described in the VEP.

That's is another great point. We do indeed need to run a PoC for VFs and NUMA alignment. I am currently trying to set it up and provide you with updates asap.

File: veps/sig-compute/115-pcie-numa-topology-awareness/pcie-numa-topology-awareness.md [hunk 17]
```diff
@@ -0,0 +1,561 @@
...
+          - deviceName: rdma/ib_device
+            name: ib1
+        resources:
+          requests:
+            memory: 8Gi
+            nvidia.com/gpu: 4
+            rdma/ib_device: 2
+```
+
+## Implementation Roadmap
```
Comment (enp0s3) [reply veps/sig-compute/115-pcie-numa-topology-awareness/pcie-numa-topology-awareness.md]: @mresvanis Hi,
I would recommend implementing all the use-cases in Alpha, leaving Beta to increase stability by adding testing coverage and addressing changes that may come from user feedback, then having one release cycle without changes, positive feedback and stable CI I would move to GA

Comment (mresvanis) [reply veps/sig-compute/115-pcie-numa-topology-awareness/pcie-numa-topology-awareness.md]: Great point, this makes a lot of sense. Introducing new use-cases in Beta seems confusing and complicates the transition from Alpha to Beta. Thank you for pointing this out, I updated the text to reflect this.

File: veps/sig-compute/115-pcie-numa-topology-awareness/pcie-numa-topology-awareness.md [hunk 18]
```diff
@@ -0,0 +1,561 @@
...
+configured. The performance impact is substantial: cross-socket communication incurs significantly
+higher latency and bandwidth costs, and frameworks cannot leverage direct device-to-device
+communication paths.
+
+Current limitations:
+- Guest VMs lack visibility into GPU and NIC NUMA placement and device locality.
+- AI frameworks cannot optimize communication topology for multi-GPU distributed model training.
+- Cross-NUMA memory traffic degrades performance despite correct NUMA-aware device allocation.
+
+## Goals
```
[the remark this answers is not in the document]
Comment (EdDev) [reply veps/sig-compute/115-pcie-numa-topology-awareness/pcie-numa-topology-awareness.md]: Thank you.
Please resolve this thread.

File: veps/sig-compute/115-pcie-numa-topology-awareness/pcie-numa-topology-awareness.md [hunk 19]
```diff
@@ -0,0 +1,561 @@
...
+- Add a new feature gate `PCINUMAAwareTopology`.
+- When the afore-mentioned feature gate and `guestMappingPassthrough` are enabled, mirror PCIe
+  host device NUMA topology in the guest.
+
+## Non Goals
+
+- This VEP focuses solely on ensuring optimal guest PCIe topology generation
+  when devices are locally available. It does not address changes to Kubernetes
+  device/resource allocation for guaranteeing NUMA-aligned device allocation.
+  Dynamic Resource Allocation (DRA) will help address that in the near future.
```
Comment (EdDev) [reply veps/sig-compute/115-pcie-numa-topology-awareness/pcie-numa-topology-awareness.md]: > KubeVirt can request devices (and other resources) for a VM/Pod

I would expect that ether Kubevirt abstracts this for the user (i.e. it asks for this through the Kubevirt API) or it needs to configure/integrate something else to make this happen.
I separate the component responsible and the actual implementation from the request method/API.

Most of it should be document, I agree. But some initial explanation here will help clarify Kubevirt part better IMO.

File: veps/sig-compute/115-pcie-numa-topology-awareness/pcie-numa-topology-awareness.md [hunk 20]
```diff
@@ -0,0 +1,561 @@
...
+- Multiple PCIe devices per NUMA node (up to 14+ devices as shown in examples).
+- Complex multi-socket systems with 4+ NUMA nodes.
+- Efficient controller allocation avoiding resource conflicts.
+
+## Update/Rollback Compatibility
+
+- Feature is opt-in via existing `spec.domain.cpu.numa.guestMappingPassthrough` flag.
+- Existing VMs without NUMA specifications continue to work unchanged.
+- Devices reporting NUMA node `-1` fallback to the default `pci.0` layout.
+- Rollback gracefully disables NUMA-aware features without affecting running workloads.
```
Comment (EdDev) [reply veps/sig-compute/115-pcie-numa-topology-awareness/pcie-numa-topology-awareness.md]: Yes, thank you.
If you can, lets call it "Fallback" to clarify this.

File: veps/sig-compute/115-pcie-numa-topology-awareness/pcie-numa-topology-awareness.md [hunk 21]
```diff
@@ -0,0 +1,625 @@
...
+- As a cloud provider, I want to offer GPU instances with predictable, high-performance characteristics.
+
+## Repos
+
+- [kubevirt](https://github.com/kubevirt/kubevirt)
+
+## Design
+
+When creating a VM with NUMA CPU passthrough enabled, KubeVirt will automatically mirror
+the host's PCIe device NUMA topology in the guest. This preserves device locality information
```
Comment (EdDev) [inline veps/sig-compute/115-pcie-numa-topology-awareness/pcie-numa-topology-awareness.md]: Starting the discussion here on the "default" operation.
An active discussion is on the root discussion page: https://github.com/kubevirt/enhancements/pull/116#issuecomment-3666959580

The decision is to have this enabled by default + reflect the existing topology.
And to allow opt-out in order to support the previous legacy behavior?
Is that the bottom line?

@vladikr , @xpivarc

File: veps/sig-compute/115-pcie-numa-topology-awareness/pcie-numa-topology-awareness.md [hunk 22]
```diff
@@ -0,0 +1,561 @@
...
+
+1. Discovers the NUMA node for each passthrough PCIe device and if it matches any of the guest's NUMA nodes.
+2. Filters out devices without NUMA alignment to be placed on the default `pci.0` bus using the existing root slot assigner.
+3. Groups devices by NUMA node: `numaNode -> []api.HostDevice`.
+4. Creates `pcie-expander-bus` controllers with `<target busNr="X"><node>Y</node></target>` elements for each NUMA node.
+5. Creates one `pcie-root-port` controller per device under the respective `pcie-expander-bus` controller.
+6. Assigns device addresses to the `pcie-root-port` aligned with its NUMA node.
+7. Inserts the controllers created into the domain before adding the host devices.
+
+### Schema Changes
```
Comment (mresvanis) [reply veps/sig-compute/115-pcie-numa-topology-awareness/pcie-numa-topology-awareness.md]: I updated the text to reflect the fact that those schema changes are in sync with the libvirt Go XML module.

**Comment by mresvanis** [comment]
> > > I like this proposal; to me, it looks excellent and thoughtful. However, I would like to avoid any VMI spec API changes for this.
> > > I believe there is only one open issue/risk related to guest ABI breakage, particularly for existing Windows VMs.
> > > What I would like to suggest is to persist the initial startup behavior in the VMI status to mitigate this without disrupting existing VMs.
> > > Specifically, On start we could add `PCITopologyMode: Legacy` for existing VMs and `PCITopologyMode: NUMAAware` for new ones, or just add a simple boolean `numaAwarePCITopologyApplied: true` for newly created ones. And then honor the stored value: rebuild the domain XML with the same topology mode.
> > > Then, for users who want to opt-out of this persistence we can introduce an annotation, something like `kubevirt.io/force-pci-numa-topology-upgrade: "true"`
> > > What do you think?
> > 
> > 
> > Chatting with @iholder101, Itamar suggested avoiding the more static vmi status api. Storing temporary or internal state in .status can feel like overkill, and the idea is that it may not be long-term. It also has only limited value to the users. At some point, most VMs will migrate to the new topology.
> > With that said we could use annotations to store this internal/temporary state: `internal.kubevirt.io/pci-topology-mode: "legacy"` And add a validating webhook to reject user and only allow controller writes (we already have a similar mechanism).
> > For users who wish to opt out ( to force the upgrade), we can still use a separate user facing annotation: `kubevirt.io/upgrade-pci-topology: "true"` Then if it's present, the controller clears or overrides the internal annotation and applies the new topology.
> 
> @mresvanis I can't find a reference to this suggestion in the VEP, maybe worth to describe it in the "Upgrade & Rollback" section.

@enp0s3 thank you for pointing this out and your suggestion! I updated the section `Update/Rollback Compatibility` with all this information. PTAL

**Comment by EdDev** [comment]
> > > Chatting with @iholder101, Itamar suggested avoiding the more static vmi status api. Storing temporary or internal state in .status can feel like overkill, and the idea is that it may not be long-term. It also has only limited value to the users. At some point, most VMs will migrate to the new topology.
> > > With that said we could use annotations to store this internal/temporary state: `internal.kubevirt.io/pci-topology-mode: "legacy"` And add a validating webhook to reject user and only allow controller writes (we already have a similar mechanism).
> > > For users who wish to opt out ( to force the upgrade), we can still use a separate user facing annotation: `kubevirt.io/upgrade-pci-topology: "true"` Then if it's present, the controller clears or overrides the internal annotation and applies the new topology.
> > 
> > 
> > @mresvanis I can't find a reference to this suggestion in the VEP, maybe worth to describe it in the "Upgrade & Rollback" section.
> 
> @enp0s3 thank you for pointing this out and your suggestion! I updated the section `Update/Rollback Compatibility` with all this information. PTAL

I think this is controversial and we do not have time for this discussion. Can we keep it in the open-questions and work on it through the Alpha stage?

**Review by orelmisan — APPROVED** [review]
Thank you @mresvanis !
Please check the boxes in the checklist :)

**Comment by vladikr** [comment]
/approve
Thank you!
